synapse/logging/opentracing.py

```diff
@@ -232,6 +233,30 @@ class _DummyTagNames(object):
     LogContextScopeManager = None  # type: ignore
 
 
+try:
+    from rust_python_jaeger_reporter import Reporter
+
+    @attr.s(slots=True, frozen=True)
+    class _WrappedRustReporter:
+        """Wrap the reporter to ensure `report_span` never throws.
+        """
+
+        _reporter = attr.ib(type=Reporter, default=lambda: Reporter())
+
+        def set_process(self, *args, **kwargs):
+            return self._reporter.set_process(*args, **kwargs)
+
+        def report_span(self, span):
+            try:
+                return self._reporter.report_span(span)
+            except Exception:
+                logger.exception("Failed to report span")
```

Comment:
Is this due to lack of confidence in the underlying `report_span` code or a just-in-case or?

Should this be in the `RustReporter`-class instead?

(Also, wouldn't this be simpler as a sub-class instead of a wrapper?)

Reply:
This is called from a lot of places, so an exception here can turn fairly fatal to the entire app, so I think an abundance of caution is probably warranted here.

We have basically three options here:

1. Make the native class not raise, which is probably mostly doable, though a) there is a layer between the python and the native code that can raise exceptions and b) means figuring out logging from the rust library (which is probably fine).
2. Move this python class into the library so that it becomes a mixed python/native library. I thought this would turn out hard, but actually looks like the tooling now supports it (some of their docs say it doesn't, but I think they're outdated)
3. What we've done here.

We should probably do a combination of 1 and 2 TBH.

It's probably also worth noting that the rust library currently does throw occasionally since a) it assumes something is around to catch and log exceptions and b) the python jaeger client is *really* inconsistent with what data types pass to the `Reporter` objects, and so there is a bunch of code in the rust lib to change types from the python custom class to the thrift defined classes, and they're not totally fool proof. (Plus, if you try and give a native library an integer bigger than 128 bits things explode).

> (Also, wouldn't this be simpler as a sub-class instead of a wrapper?)

Maybe? I'm not a super fan of sub classing and I'm not 100% if there are any gotchas with subclassing native types.

Reply:
> We should probably do a combination of 1 and 2 TBH.

We could make this improvement in the future, I suppose. 👍

Reply:
Is there anything we should log here to help identify what happened to cause the exception? Or is using `logger.exception` enough to provide what we need?

Reply:
Logging the exception should tell us everything bar what `span` is, and sentry should tell us that.